src/test/admin_socket.cc

```diff
@@ -328,6 +330,175 @@ TEST(AdminSocket, bind_and_listen) {
+    }
+    if (after) {
+      f.dump_float("after", *after);
+    }
+    if (cancel) {
+      f.dump_bool("cancel", true);
+    }
+    f.close_section();
+
+    bufferlist command;
+    f.flush(command);
+
+    std::string response;
+
+    asock_client->do_request(command.to_str(), &response);
+    return response;
+  }
+
+  std::string send_raise_cancel(std::optional<std::string> arg = std::nullopt) {
+    return send_raise(arg, std::nullopt, true);
+  }
+
+  std::string send_raise(std::string arg, std::optional<double> after = std::nullopt) {
+    return send_raise(arg, after, false);
+  }
+};
+
+std::atomic<int> AdminSocketRaise::count_usr1 = {};
+std::atomic<int> AdminSocketRaise::count_usr2 = {};
+
+TEST_F(AdminSocketRaise, List) {
+  auto r = send_raise("-l");
+  json_spirit::mValue v;
+  ASSERT_TRUE(json_spirit::read(r, v));
+  ASSERT_EQ(json_spirit::Value_type::obj_type, v.type());
+  EXPECT_EQ(SIGUSR1, v.get_obj()["USR1"].get_int());
+  EXPECT_EQ(SIGUSR2, v.get_obj()["USR2"].get_int());
+}
+
+TEST_F(AdminSocketRaise, ImmediateFormats) {
+  send_raise(fmt::format("-{}", SIGUSR1));
+  send_raise("uSr1");
+  send_raise("-UsR2");
+  send_raise(fmt::format("{}", SIGUSR2));
+  EXPECT_EQ(2, count_usr1.load());
+  EXPECT_EQ(2, count_usr2.load());
+}
+
+TEST_F(AdminSocketRaise, Async)
+{
+  using std::chrono::milliseconds;
+
+  ASSERT_EQ("", send_raise(fmt::format("{}", SIGUSR1)));
+  ASSERT_EQ("", send_raise("USR2", 0.1));
+
+  EXPECT_EQ(1, count_usr1.load());
+  EXPECT_EQ(0, count_usr2.load());
+
+  this_thread::sleep_for(milliseconds(150));
+
+  EXPECT_EQ(1, count_usr1.load());
+  EXPECT_EQ(1, count_usr2.load());
+}
+
+TEST_F(AdminSocketRaise, AsyncReschedule)
+{
+  using std::chrono::milliseconds;
+
+  ASSERT_EQ("", send_raise("USR1", 0.1));
+  ASSERT_EQ("", send_raise("USR2", 0.2));
+
+  EXPECT_EQ(0, count_usr1.load());
+  EXPECT_EQ(0, count_usr2.load());
+
+  this_thread::sleep_for(milliseconds(150));
+
+  // USR1 got overridden by the second async schedule
+  EXPECT_EQ(0, count_usr1.load());
+  EXPECT_EQ(0, count_usr2.load());
+
+  this_thread::sleep_for(milliseconds(100));
+  EXPECT_EQ(0, count_usr1.load());
+  EXPECT_EQ(1, count_usr2.load());
+}
+
+TEST_F(AdminSocketRaise, AsyncCancel)
+{
+  using std::chrono::milliseconds;
+
+  ASSERT_EQ("", send_raise("USR1", 0.1));
+
+  EXPECT_EQ(0, count_usr1.load());
+  EXPECT_EQ(0, count_usr2.load());
+
+  ASSERT_EQ("", send_raise_cancel("USR2"));
+
+  this_thread::sleep_for(milliseconds(150));
+
+  // cancel shouldn't have worked because the signals
+  // didn't match
+  EXPECT_EQ(1, count_usr1.load());
+
+  ASSERT_EQ("", send_raise("USR2", 0.1));
+  ASSERT_EQ("", send_raise_cancel("USR2"));
+
+  this_thread::sleep_for(milliseconds(150));
+
+  // cancel must have worked
+  EXPECT_EQ(0, count_usr2.load());
+
+  ASSERT_EQ("", send_raise("USR1", 0.1));
+  ASSERT_EQ("", send_raise_cancel());
+
+  // cancel must have worked, the counter stays 1
+  EXPECT_EQ(1, count_usr1.load());
+}
+
```

Comment:
Add a test for `--after 0.1 --signal SIGCONT` then `--signal SIGSTOP`.

Reply:
I thought that the existing urs1 and usr2 tests were sufficient, they cover exactly this scenario logically. But I can add such test, yes.

Reply:
done

Reply:
Well this is the basis for the feature + it validates that the daemon can resume itself (which is why we have all the fork complexity to begin with!).